File: app/src/main/java/com/wp/csmu/classschedule/network/NetWorkHelper.java
package com.wp.csmu.classschedule.network;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import com.wp.csmu.classschedule.application.MyApplication;
import com.wp.csmu.classschedule.view.bean.Score;
import com.wp.csmu.classschedule.view.scheduletable.Subjects;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import okhttp3.FormBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class NetWorkHelper {
    public static String token = "";
    public static int weekIndex = 0;
    public static String endTime = "";
    public static String startTime = "";
    public static String termName = "";
    public static List<Subjects> subjects = new ArrayList<>();
    private static OkHttpClient client = new OkHttpClient();

    public static void login(String account, String password) throws Exception {
        String url = "http://jiaowu.csmu.edu.cn:8099/app.do?method=authUser&xh=" + account + "&pwd=" + password;
        Request request = new Request.Builder().url(url).get().build();
        Response response = client.newCall(request).execute();
        JSONObject jsonObject = new JSONObject(response.body().string());
        if (jsonObject.getBoolean("success")) {
            token = jsonObject.getString("token");
            SharedPreferences sharedPreferences = MyApplication.getContext().getSharedPreferences("user", Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putString("account", account);
            editor.putString("password", password);
            editor.putString("token", token);
            editor.putString("userName", jsonObject.getString("userrealname"));
            editor.putString("userCollege", jsonObject.getString("userdwmc"));
            editor.apply();
        } else {
            throw new IllegalStateException("用户名或密码错误");
        }
    }

    private static int getCurrentWeek() throws Exception {
        Date date = new Date();
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd");
        String url = "http://jiaowu.csmu.edu.cn:8099/app.do?method=getCurrentTime&currDate=" + simpleDateFormat.format(date);
        Request request = new Request.Builder().url(url).addHeader("token", token).get().build();
        Response response = client.newCall(request).execute();
        JSONObject jsonObject = new JSONObject(response.body().string());
        try {
            weekIndex = jsonObject.getInt("zc");
            endTime = jsonObject.getString("e_time");
            startTime = jsonObject.getString("s_time");
            termName = jsonObject.getString("xnxqh");
        } catch (JSONException e) {
            Log.i("getCurrentWeek", "getCurrentWeek: " + e.getMessage());
            if ("-1".equals(jsonObject.getString("token"))) {
                SharedPreferences sharedPreferences = MyApplication.getContext().getSharedPreferences("user", Context.MODE_PRIVATE);
                String account = sharedPreferences.getString("account", null);
                String password = sharedPreferences.getString("password", null);
                if (account != null && password != null) {
                    login(account, password);
                    getCurrentWeek();
                }
            }
        }
        return weekIndex;
    }

    public synchronized static void getSchedule(String account, int weekIndex) throws Exception {
        subjects.clear();
        NetWorkHelper.weekIndex = ((weekIndex == -1) ? getCurrentWeek() : weekIndex);
        String url = "http://jiaowu.csmu.edu.cn:8099/app.do?method=getKbcxAzc&xh=" + account + "&xnxqid=" + termName + "&zc=" + NetWorkHelper.weekIndex;
        Request request = new Request.Builder().url(url).addHeader("token", token).get().build();
        Response response = client.newCall(request).execute();
        JSONArray jsonArray = new JSONArray(response.body().string());
        for (int i = 0; i < jsonArray.length(); i++) {
            JSONObject jsonObject = jsonArray.getJSONObject(i);
            String teacher = jsonObject.getString("jsxm");
            String room = jsonObject.getString("jsmc");
            String time = jsonObject.getString("kcsj");
            String name = jsonObject.getString("kcmc");
            int day = Integer.valueOf(time.substring(0, 1));
            int step = time.substring(1).length() / 2;
            int start = Integer.valueOf(time.substring(1, 3));
            Subjects subjects = new Subjects();
            subjects.setDay(day);
            subjects.setName(name);
            subjects.setRoom(room);
            subjects.setStart(start);
            subjects.setStep(step);
            subjects.setTeacher(teacher);

            NetWorkHelper.subjects.add(subjects);
        }
    }

    public synchronized static String[] getHomepage(String account) throws Exception {
        if (token.equals("")){
            SharedPreferences sharedPreferences=MyApplication.getContext().getSharedPreferences("user",Context.MODE_PRIVATE);
            login(account,sharedPreferences.getString("password",""));
        }
        String[] info = new String[12];
        FormBody formBody = new FormBody.Builder().add("userAccountType", "2").build();
        Request request = new Request.Builder().url("http://jiaowu.csmu.edu.cn:8099/app.do?method=getUserInfo&xh=" + account)
                .addHeader("token", token).post(formBody).build();
        Response response = client.newCall(request).execute();
        try {
            JSONObject jsonObject = new JSONObject(response.body().string());
            info[0] = jsonObject.getString("xm");
            info[1] = jsonObject.getString("xh");
            info[2] = jsonObject.getString("xb");
            info[3] = jsonObject.getString("dh");
            info[4] = jsonObject.getString("qq");
            info[5] = jsonObject.getString("email");
            info[6] = jsonObject.getString("yxmc");
            info[7] = jsonObject.getString("zymc");
            info[8] = jsonObject.getString("bj");
            info[9] = jsonObject.getString("rxnf");
            info[10] = jsonObject.getString("nj");
            info[11] = jsonObject.getString("fxzy");
        }catch (JSONException e){
            if (e.toString().contains("no value for")){
                SharedPreferences sharedPreferences=MyApplication.getContext().getSharedPreferences("user",Context.MODE_PRIVATE);
                login(account,sharedPreferences.getString("password",""));
                info=getHomepage(account);
            }
        }
        return info;
    }

    public synchronized static List<Score> getScore(String account,String selectedTerm) throws Exception {
        if (token.equals("")){
            SharedPreferences sharedPreferences=MyApplication.getContext().getSharedPreferences("user",Context.MODE_PRIVATE);
            login(account,sharedPreferences.getString("password",""));
        }
        List<Score>list=new ArrayList<>();
        Request request=new Request.Builder().url("http://jiaowu.csmu.edu.cn:8099/app.do?method=getCjcx&xh="+account+"&xnxqid="+selectedTerm)
                .addHeader("token",token).get().build();
        Response response=client.newCall(request).execute();
        JSONObject jsonObject=new JSONObject(response.body().string());
        JSONArray jsonArray=jsonObject.getJSONArray("result");
        try {
            for(int i=0;i<jsonArray.length();i++){
                JSONObject object=jsonArray.getJSONObject(i);
                Score score=new Score();
                score.setScore(object.getString("zcj"));
                score.setCredit(object.getInt("xf"));
                score.setSubjectEnglish(object.getString("kcywmc"));
                score.setSubjectAttribute(object.getString("kclbmc"));
                score.setTerm(object.getString("xqmc"));
                score.setSubject(object.getString("kcmc"));
                score.setExamAttribute(object.getString("ksxzmc"));
                score.setSubjectNature(object.getString("kcxzmc"));
                score.setNote(object.getString("bz"));
                list.add(score);
            }
        }catch (JSONException e){
            if (e.toString().contains("no value for")){
                SharedPreferences sharedPreferences=MyApplication.getContext().getSharedPreferences("user",Context.MODE_PRIVATE);
                login(account,sharedPreferences.getString("password",""));
               list=getScore(account,selectedTerm);
            }
        }
        return list;
    }

    public static JSONArray getTerms(String account) throws Exception {
        Request request=new Request.Builder().url("http://jiaowu.csmu.edu.cn:8099/app.do?method=getXnxq&xh="+account)
                .addHeader("token",token).get().build();
        Response response=client.newCall(request).execute();
        return new JSONArray(response.body().string());
    }
}

File: app/src/main/java/com/wp/csmu/classschedule/network/NetworkHelper.kt
package com.wp.csmu.classschedule.network

import android.content.Context
import android.util.Base64
import android.util.Log
import com.wp.csmu.classschedule.activity.ScoreActivity
import com.wp.csmu.classschedule.application.MyApplication
import com.wp.csmu.classschedule.view.bean.Score
import com.wp.csmu.classschedule.view.scheduletable.Subjects
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.util.regex.Pattern

object NetworkHelper {
    private val client = OkHttpClient()
    var cookie = ""
    var termSelection = HashMap<String, String>()
    @JvmStatic
    var schedules = HashSet<Subjects>()

    @JvmStatic
    var termBeginsTime = ""

    object Login {
        @Throws(Exception::class)
        internal fun login(username: String, password: String) {
            if (cookie == "") {
                getCookie()
            }
            val url = "http://jiaowu.csmu.edu.cn:8099/jsxsd/xk/LoginToXk"
            val formBody = FormBody.Builder().add("encoded", encryptPassword(username, password)).build()
            val request = Request.Builder().url(url).addHeader("Cookie", cookie).post(formBody).build()
            val response = client.newCall(request).execute()
            val document = Jsoup.parse(response.body()!!.string())
            if (document.title() == "学生个人中心") {
                println("登录成功")
            } else {
                Log.i("NetworkHelper", document.title())
                throw Exception("登录失败")
            }
        }

        private fun encryptPassword(username: String, password: String): String {
            return String(Base64.encode(username.toByteArray(), Base64.DEFAULT)).plus("%%%")
                    .plus(String(Base64.encode(password.toByteArray(), Base64.DEFAULT)))
        }

        internal fun getCookie() {
            val url = "http://jiaowu.csmu.edu.cn:8099/jsxsd/"
            val request = Request.Builder().url(url).get().build()
            val response = client.newCall(request).execute()
            cookie = response.header("Set-Cookie")!!
        }
    }

    object GetSchedule {
        @Throws(Exception::class)
        @JvmStatic
        fun getSchedule(account: String, password: String) {
            NetworkHelper.Login.login(account, password)
            getSchedulePage()
        }

        private fun getSchedulePage() {
            val url = "http://jiaowu.csmu.edu.cn:8099/jsxsd/xskb/xskb_list.do"
            //val formBody = FormBody.Builder().add("xnxq01id", "2018-2019-1").build()
            val request = Request.Builder().url(url).header("Cookie", cookie).get().build()
            val response = client.newCall(request).execute()
            val document = Jsoup.parse(response.body()!!.string())
            /*********获取学期选择************/
            val termDocument = document.select("select[name=xnxq01id]").first().select("option")
            for (element in termDocument) {
                val key = element.attr("value")
                val value = element.text()
                termSelection[key] = value
                //println("key=$key map=$value")
            }
            /********获取课程表***************/
            val scheduleDocument = document.select("table[id=kbtable]").first()
            getSchedule(scheduleDocument)
            /********获取开学时间***************/
            termBeginsTime = NetworkHelper.GetTermBeginsTime.mGetTermBeginsTime()
        }

        private fun getSchedule(htmlTable: Element) {
            val elements = htmlTable.selectFirst("tbody").select("tr")
            for (i in 1 until elements.size) {
                val classElements = elements[i].select("td").select("div[class=kbcontent]")
                for (j in 0 until classElements.size) {
                    parseSchedule(classElements[j], j + 1)
                }
            }
        }

        private fun parseSchedule(element: Element, currentDay: Int) {
            val string = element.toString().replace(Regex("-{3,}"), "</div><div>")
            val document = Jsoup.parse(string)
            val elements = document.select("div")
            for (element1 in elements) {
                val className = element1.ownText()
                //无教师时
                var teacherName = ""
                try {
                    teacherName = element1.selectFirst("font[title=老师]").text()
                } catch (e: java.lang.NullPointerException) {
                }
                //无时间时
                var weeks: Array<Int>? = null
                var startTime = 0
                var endTime = 0
                try {
                    val string2 = element1.selectFirst("font[title=周次(节次)]").text()
                    weeks = parseWeeks(string2)!!
                    val times = parseTimes(string2)!!
                    startTime = times[0]
                    endTime = times[1]
                } catch (e: java.lang.NullPointerException) {
                }
                //无教室时
                var classRoom = ""
                try {
                    classRoom = element1.selectFirst("font[title=教室]").text()
                } catch (e: java.lang.NullPointerException) {
                }
                //无课程时
                try {
                    val subjects = Subjects()
                    subjects.day = currentDay
                    subjects.name = className
                    subjects.room = classRoom
                    subjects.start = startTime
                    subjects.step = endTime - startTime + 1
                    subjects.teacher = teacherName
                    subjects.weeks = weeks!!.toMutableList()
                    subjects.end = endTime
                    schedules.add(subjects)
                } catch (e: Exception) {
                }
            }
        }

        private fun parseWeeks(text: String): Array<Int>? {
            val pattern = Pattern.compile("(.*)\\(周\\).*")
            val matcher = pattern.matcher(text)
            if (matcher.find()) {
                val array = matcher.group(1).split(",")
                val list = ArrayList<Int>()
                for (string in array) {
                    if (string.contains("-")) {
                        val array2 = string.split("-")
                        for (i in array2[0].toInt()..array2.last().toInt()) {
                            list.add(i)
                        }
                    } else {
                        list.add(string.toInt())
                    }
                }
                return list.toArray(Array(list.size) { 0 })
            }
            return null
        }

        private fun parseTimes(text: String): Array<Int>? {
            val pattern = Pattern.compile(".*\\[(.*)节]")
            val matcher = pattern.matcher(text)
            if (matcher.find()) {
                val string = matcher.group(1)
                return if (string.length == 2) {
                    arrayOf(string.toInt(), string.toInt())
                } else {
                    val array = string.split("-")
                    arrayOf(array[0].toInt(), array.last().toInt())
                }
            }
            return null
        }
    }

    object GetTermBeginsTime {
        private fun getTermBeginsPage(): String {
            val document = Jsoup.connect("http://jiaowu.csmu.edu.cn:8099/jsxsd/jxzl/jxzl_query").header("Cookie", cookie).get()
            val table = document.select("table[id=kbtable] > tbody")[0]
            return parseTermBeginsTime(table)
        }

        internal fun mGetTermBeginsTime(): String = getTermBeginsPage()

        private fun parseTermBeginsTime(element: Element): String {
            val trs = element.select("tr")
            for (i in 1 until trs.size) {
                val tds = trs[i].select("td")
                for (j in 0 until tds.size) {
                    if (tds[j].attr("title") === null || tds[j].attr("title").trim() == "") {

                    } else {
                        return tds[j].attr("title").replace("年", "-").replace("月", "-")
                    }
                }
            }
            return ""
        }
    }

    object GetGrades {
        @Throws(Exception::class)
        @JvmStatic
        public fun getGrades(term: String): ArrayList<Score> {
            val url = "http://jiaowu.csmu.edu.cn:8099/jsxsd/kscj/cjcx_list"
            val formBody = FormBody.Builder().add("kksj", term).add("kcxz", "").add("kcmc", "").add("xsfs", "all").build()
            val request = Request.Builder().url(url).addHeader("Cookie", cookie).post(formBody).build()
            val document = client.newCall(request).execute().body()!!.string()
            return parseGrades(document)
        }

        //传入学期名和ID，会修改内容
        //返回目前默认选择的学期编号
        @Throws(Exception::class)
        @JvmStatic
        public fun queryTerms(scoreActivity: ScoreActivity): Int {
            val sharedPreferences = MyApplication.getContext().getSharedPreferences("user", Context.MODE_PRIVATE)
            val username = sharedPreferences.getString("account", "")
            val password = sharedPreferences.getString("password", "")
            Login.login(username, password)
            val url = "http://jiaowu.csmu.edu.cn:8099/jsxsd/kscj/cjcx_query"
            val request = Request.Builder().url(url).addHeader("Cookie", cookie).get().build()
            val document = client.newCall(request).execute().body()!!.string()
            val termDocument = Jsoup.parse(document).select("select[id=kksj]").select("option")
            var selected = 0
            for ((i, element) in termDocument.withIndex()) {
                scoreActivity.termName.add(element.text())
                scoreActivity.termId.add(element.attr("value"))
                if (element.hasAttr("selected")) {
                    selected = i
                }
            }
            return selected
        }

        private fun parseGrades(document: String): ArrayList<Score> {
            val document = Jsoup.parse(document)
            val table = document.select("table[id=dataList] > tbody")
            val trs = table.select("tr")

            if (trs.size == 2 && trs[1].text() == "未查询到数据") {
                return ArrayList()
            }
            val scores = mutableListOf<Score>() as ArrayList
            for (i in 1 until trs.size) {
                val tds = trs[i].select("td")
                val term = tds[1].text()
                val name = tds[3].text()
                val score = tds[4].text()
                val skillScores = tds[5].text()
                val performanceScore = tds[6].text()
                val knowledgePoints = tds[7].text()
                val credits = tds[9].text().toDouble()
                val examAttribute = tds[10].text()
                val subjectNature = tds[14].text()
                val subjectAttribute = tds[15].text()
                scores.add(Score(term, name, score, skillScores, performanceScore, knowledgePoints, credits, examAttribute, subjectNature, subjectAttribute))
            }
            return scores
        }
    }
}